senders/client_factory.go

```diff
@@ -85,9 +93,80 @@ func NewSender(wfURL string, setters ...Option) (Sender, error) {
 		return nil, fmt.Errorf("unable to create sender config: %s", err)
 	}
 
+	var tokenService = token.NewNoopTokenService()
+	if len(cfg.Token) > 0 {
+		tokenService = token.NewWavefrontTokenService(cfg.Token)
+	} else if len(cfg.CSPBaseUrl) > 0 && len(cfg.CSPClientId) > 0 && len(cfg.CSPClientSecret) > 0 {
+		tokenService = token.NewCSPServerToServerTokenService(cfg.CSPBaseUrl, cfg.CSPClientId, cfg.CSPClientSecret)
+	}
+
+	switch tokenService.(type) {
+	case *token.CspServerToServerTokenService:
+		log.Println("The Wavefront SDK will use CSP authentication when communicating with the Wavefront Backend for Direct Ingestion.")
+	case token.WavefrontTokenService:
+		log.Println("The Wavefront SDK will use an API TOKEN when communicating with the Wavefront Backend for Direct Ingestion.")
+	case token.TokenNoOpService:
+		log.Println("The Wavefront SDK will communicate with a Wavefront Proxy.")
+	}
+
+
+	client := internal.NewClient(cfg.Timeout, cfg.TLSConfig)
+	metricsReporter := internal.NewReporter(cfg.metricsURL(), tokenService, client)
+	tracesReporter := internal.NewReporter(cfg.tracesURL(), tokenService, client)
+
+	sender := &wavefrontSender{
+		defaultSource: internal.GetHostname("wavefront_direct_sender"),
+		proxy:         !cfg.Direct(),
+	}
+	if cfg.SendInternalMetrics {
+		sender.internalRegistry = sender.realInternalRegistry(cfg)
+	} else {
+		sender.internalRegistry = sdkmetrics.NewNoOpRegistry()
+	}
+	sender.pointHandler = newLineHandler(metricsReporter, cfg, internal.MetricFormat, "points", sender.internalRegistry)
+	sender.histoHandler = newLineHandler(metricsReporter, cfg, internal.HistogramFormat, "histograms", sender.internalRegistry)
+	sender.spanHandler = newLineHandler(tracesReporter, cfg, internal.TraceFormat, "spans", sender.internalRegistry)
+	sender.spanLogHandler = newLineHandler(tracesReporter, cfg, internal.SpanLogsFormat, "span_logs", sender.internalRegistry)
+	sender.eventHandler = newLineHandler(metricsReporter, cfg, internal.EventFormat, "events", sender.internalRegistry)
+
+	sender.Start()
+	return sender, nil
+}
+
+func CSPBaseUrl(CSPBaseUrl string) Option {
+	return func(cfg *configuration) {
+		cfg.CSPBaseUrl = CSPBaseUrl
+	}
+}
+
+// NewCSPSender creates Wavefront client that uses CSP auth
+func NewCSPSender(wfURL string, cspClientId string, cspClientSecret string, setters ...Option) (Sender, error) {
```

Comment:
Let's find a way to collapse this into just one NewSender function. CSP tokens can be passed as an Option.